lib/gyro/liquidgen/generator.rb

```diff
@@ -0,0 +1,120 @@
+# Copyright 2016 - Niji
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'gyro/xcdatamodel/parser'
+require 'gyro/liquidgen/whitespace_patch'
+require 'liquid'
+require 'pathname'
+
+module Gyro
+  module Liquidgen
+    # Generates arbitrary output from the input datamodel, using a Liquid template provided by the user
+    #
+    class Generator
+      # INCLUDES #############################################################
+
+      include Gyro::XCDataModel::Parser
+
+      # PUBLIC METHODS #######################################################
+
+      # Declare some custom Liquid Filters used by the template, then render it
+      module CustomFilters
+        def escape_quotes(input)
+          input.gsub('"', '\"')
+        end
+
+        def snake_to_camel_case(input)
+          input.split('_').map(&:capitalize).join
+        end
+
+        def uncapitalize(input)
+          input_strip = input.strip
+          input_strip[0, 1].downcase + input_strip[1..-1]
+        end
+      end
+
+      def initialize(xcdatamodel, template_dir, output_dir, params)
```

Comment:
Thinking about it, it was a bad practice having a constructor which actually does some side-effects and actions (here actually generating files on disk).

Maybe it would be better to separate initialisation and actions/side-effects, and have:

* `def initialize(template_dir, output_dir, params)` which would only use the parameters to store values in `@root_template`, `@file_template`, `@output_dir` and `@params` properties
* then have a `def generate(xcdatamodel)` function which does the actual `xcdatamodel.to_h['entities'].each do …` loop generating the files on disk?

Then adapt `bin/gyro` accordingly to replace the `Gyro::Liquid::Generator.new(…)` with a adjusted call `gen = Gyro::Liquid::Generator.new(…)` + calling `gen.generate(xcdatamodel)`?